_posts/2020-04-11-native-mobile-to-flutter.markdown

```diff
@@ -0,0 +1,122 @@
+
+### Choosing a State Management pattern that matched our needs.
+
+There were several state management architectures to choose from, but we opted for simplicity, the [Scoped Model pattern](https://github.com/brianegan/scoped_model).  The scoped model pattern is a simple change notifier pattern bound to Flutter's UI widget hierarchy.  It has gradually been superceded in the community by the [Provider library](https://github.com/rrousselGit/provide) which is based on Scoped Model.  We will most likely be migrating to the provider library within the next year.
+
+We use this pattern with our SQLite database to allow sql table subscriptions that trigger redraws on subscribed user interfaces when data is modified on disk.  We also use a very simple stream subscription pattern to propagate state changes that are not tied to database state.  Flutter provides the excellent [StreamBuilder class](https://api.flutter.dev/flutter/widgets/StreamBuilder-class.html) to facilitate state propagation.
+
+We evaluated a few other patterns mentioned in flutter resources, particularly Redux and BLoC, but chose the simplicity of Scoped Model over the granularity and associated complexity that the others presented.  
+
+### Native integrations. 
+
+Flutter offers a very simple message passing api for communicating with native code.  We needed to use native integrations for a relatively small group of features: 
+
+- Push notification.
+- Deep link handling.
+- Connectivity detection.
+- Security features.
+- Device information. 
+
+_This platform specific code amounted to less than 2% of our total lines of code._ 
+
+We were able to write this functionality easily and communicate between Flutter and native with very little code overhead. Flutter provides a [very simple api](https://flutter.dev/docs/development/platform-integration/platform-channels#architecture) for communicating bidirectionally between native and dart code.
+
+### Unit tests.
+
+Unit testing in Flutter is straightforward, but lacks the robust testing utilities and third party libraries found in React.  We rolled most of our own testing scaffolding aside from the simple test runner offered by flutter.
+
+### UI testing. 
+
+UI testing in Flutter only runs in a simulator/emulator, a pretty significant drawback.  Though we were able to write effective widget tests, we found them to be brittle.  This is definitely an area that needs improvements in future iterations of Flutter.  We opted to keep the majority of presentation logic in view models so that we could test the presentation layer with unit tests wherever possible rather than employing widget tests.
+
+### Continuous integration.
+
+We opted for CodeMagic as our CI service.  It was the only commercial service we found that had specialized flutter workflows that required no additional scrpting or configuration.  We were able to integrate our git workflow with CodeMagic in under an hour.   
+
+### Security.
```

Comment:
I have a similar thought here where this would make a great separate blog post.

Reply:
With this one - we don't have many more data points.  Both of our native apps had security flaws which pen testing exposed.  Our flutter app had no security vulnerabilities exposed by similar pen testing.  This is just a great nugget for someone interested in Flutter, that a vanilla build using core flutter/dart libraries and expressing fairly normal app functionality, is secure out of the box.  Knowing that someone else out there went and paid for pen testing and got a green light on their Flutter application is a nice security blanket.   Maybe expounding to add some of what i've written in this comment could be a good value add.

Reply:
Ah I should have given more context w.r.t. to the comment. I was thinking there were a few things that could be interesting to write about (in a separate post or posts), like: 

- fork sqllite to get the encrypted data store to work
- magic login links
- Experience working with 3rd party pen testers, how we chose a company to work with, what the process is like etc. 

Stuff like that.